### PR TITLE
Expanding decoding response for transactions

### PR DESCRIPTION
**Pull request by owl352**
# Issue
After #236, we need additional data for decoding response
Also Alexey requests some changes for response
After #313 we have error in search query

# Things done
Fixed transactions error in query and view
Added new fields for decode
Updated README.md
Updated `unit.spec.js`
Added new Enums
Tiny improvments
Updated https://github.com/owl352/wasm-dpp
Added new env for network type

### CODE REVIEW

File: packages/api/README.md [hunk 1]
```diff
@@ -58,6 +58,7 @@ Reference:
 * [Documents by Data Contract](#documents-by-data-contract)
 * [Identity by Identifier](#identity-by-identifier)
 * [Identity by DPNS](#identity-by-dpns)
+* [Identity Withdrawals](#identity-withdrawals)
```
Comment (pshenmic) [inline packages/api/README.md:61]: Apply documentation for decodeStateTransition

File: packages/api/src/utils.js [hunk 2]
```diff
@@ -33,23 +35,48 @@ const decodeStateTransition = async (client, base64) => {
 
   switch (decoded.type) {
     case StateTransitionEnum.DATA_CONTRACT_CREATE: {
+      const dataContractConfig = stateTransition.getDataContract().getConfig()
+
+      decoded.internalConfig = {
+        canBeDeleted: dataContractConfig.canBeDeleted ?? null,
```
Comment (pshenmic) [inline packages/api/src/utils.js:41]: only requires_identity_encryption_bounded_key and requires_identity_decryption_bounded_key are optional

File: packages/api/src/DAPI.js [hunk 3]
```diff
@@ -51,7 +51,7 @@ class DAPI {
    * @param {string} indexName
    * @param {number} resultType
    * @param {Array<Buffer>} indexValuesList
-   * @param {Buffer} [startAtIdentifierInfo]
+   * @param {Object} [startAtIdentifierInfo]
```
Comment (pshenmic) [inline packages/api/src/DAPI.js:54]: Why is that?

File: packages/api/README.md (hunk 1 above)
Comment (pshenmic) [reply packages/api/README.md:61]: Also mention null result set for identities withdrawals

**Review by pshenmic — CHANGES_REQUESTED** [review]

File: packages/api/src/DAPI.js (hunk 3 above)
Comment (owl352) [reply packages/api/src/DAPI.js:54]: Mistake in previous PR

Comment (owl352) [reply packages/api/src/DAPI.js:54]: Documentation say it's must be object

Comment (owl352) [reply packages/api/src/DAPI.js:54]: start_at_vote_poll_id_info | Object | No | Start poll ID information for pagination

File: packages/api/src/utils.js (hunk 2 above)
Comment (owl352) [reply packages/api/src/utils.js]: ✅

File: packages/api/README.md (hunk 1 above)
Comment (owl352) [reply packages/api/README.md:61]: ✅

File: packages/api/README.md [hunk 4]
```diff
@@ -951,3 +962,76 @@ Response codes:
...
+
+```
+POST /transaction/decode
+{
+    "base64": "AAAA56Y/VzBp5vlrJR8JRCPSDLlaZjngwyM50w8dQAmAe3EAAAAAAAEBAAABYpzp8+tOQ8j6k24W7FXjqo7zZmMZcybMIDLw7VfLT0EAAQZsYWJsZXIWBBIEdHlwZRIGb2JqZWN0Egpwcm9wZXJ0aWVzFgISCmNvbnRyYWN0SWQWBBIEdHlwZRIGc3RyaW5nEgltaW5MZW5ndGgDVhIJbWF4TGVuZ3RoA1gSCHBvc2l0aW9uAwASCXNob3J0TmFtZRYEEgR0eXBlEgZzdHJpbmcSCW1heExlbmd0aANAEgltaW5MZW5ndGgDBhIIcG9zaXRpb24DAhIIcmVxdWlyZWQVAhIJc2hvcnROYW1lEgpjb250cmFjdElkEhRhZGRpdGlvbmFsUHJvcGVydGllcxMACgACQR8AOrSAQ3S/emVWILS8WyHcMA97CtY5rH7dB4DSjAm/0x6DZdZcm8jyGIdIuuTUALR8/N724YhxwhOQHqUm5ipN"
+}
+
+Response:
+{
+    "type": 0,
```
Comment (pshenmic) [inline packages/api/README.md:990]: Show all transaction types

File: packages/api/src/DAPI.js (hunk 3 above)
Comment (pshenmic) [reply packages/api/src/DAPI.js:54]: What is the object? What type is it?

**Review by pshenmic — CHANGES_REQUESTED** [review]

File: packages/api/src/utils.js [hunk 5]
```diff
@@ -33,23 +35,48 @@ const decodeStateTransition = async (client, base64) => {
...
+      decoded.publicKeyId = stateTransition.getSignaturePublicKeyId()
+      decoded.ownerId = stateTransition.getOwnerId().toString()
+      decoded.raw = stateTransition.toBuffer().toString('hex')
+
       break
     }
     case StateTransitionEnum.IDENTITY_CREATE: {
       decoded.identityId = stateTransition.getIdentityId().toString()
+      decoded.signature = stateTransition.getSignature()?.toString('hex') ?? null
+      decoded.raw = stateTransition.toBuffer().toString('hex')
```
Comment (pshenmic) [inline packages/api/src/utils.js:79]: Missing public keys for identity create

File: packages/api/README.md (hunk 1 above)
Comment (pshenmic) [inline packages/api/README.md:61]: Add decode transaction in the list

File: packages/api/src/utils.js [hunk 6]
```diff
@@ -59,26 +86,54 @@ const decodeStateTransition = async (client, base64) => {
 
       decoded.identityId = stateTransition.getIdentityId().toString()
       decoded.amount = output.satoshis * 1000
+      decoded.signature = stateTransition.getSignature()?.toString('hex') ?? null
+      decoded.raw = stateTransition.toBuffer().toString('hex')
```
Comment (pshenmic) [inline packages/api/src/utils.js:90]: Missing assetlockproof and funding address (input) data

File: packages/api/src/utils.js [hunk 7]
```diff
@@ -59,26 +86,54 @@ const decodeStateTransition = async (client, base64) => {
...
+        requiresIdentityDecryptionBoundedKey: dataContractConfig.requiresIdentityDecryptionBoundedKey ?? null,
+        requiresIdentityEncryptionBoundedKey: dataContractConfig.requiresIdentityEncryptionBoundedKey ?? null
+      }
+
+      decoded.ownerId = stateTransition.getDataContract().getOwnerId().toString()
       decoded.dataContractId = stateTransition.getDataContract().getId().toString()
+      decoded.schema = stateTransition.getDataContract().getDocumentSchemas()
       decoded.version = stateTransition.getDataContract().getVersion()
+      decoded.dataContractOwner = stateTransition.getDataContract().getOwnerId().toString()
+      decoded.raw = stateTransition.toBuffer().toString('hex')
```
Comment (pshenmic) [inline packages/api/src/utils.js:113]: Missing identity contract nonce

File: packages/api/src/utils.js [hunk 8]
```diff
@@ -33,23 +35,48 @@ const decodeStateTransition = async (client, base64) => {
...
+        requiresIdentityEncryptionBoundedKey: dataContractConfig.requiresIdentityEncryptionBoundedKey ?? null
+      }
+
       decoded.dataContractId = stateTransition.getDataContract().getId().toString()
-      decoded.identityId = stateTransition.getOwnerId().toString()
+      decoded.ownerId = stateTransition.getOwnerId().toString()
+      decoded.schema = stateTransition.getDataContract().getDocumentSchemas()
+      decoded.signature = Buffer.from(stateTransition.toObject().signature).toString('hex')
+      decoded.signatureKeyId = stateTransition.toObject().signaturePublicKeyId
+      decoded.raw = stateTransition.toBuffer().toString('hex')
```
Comment (pshenmic) [inline packages/api/src/utils.js:56]: Identity Nonce missing

Comment (pshenmic) [reply packages/api/src/utils.js:56]: User fee Increase is missing

File: packages/api/src/utils.js [hunk 9]
```diff
@@ -33,23 +35,48 @@ const decodeStateTransition = async (client, base64) => {
...
-        revision: documentTransition.getRevision()
+        revision: documentTransition.getRevision(),
+        type: documentTransition.getType(),
+        action: DocumentActionEnum[documentTransition.getAction()]
       }))
 
+      decoded.signature = Buffer.from(stateTransition.getSignature()).toString('hex')
+      decoded.publicKeyId = stateTransition.getSignaturePublicKeyId()
+      decoded.ownerId = stateTransition.getOwnerId().toString()
+      decoded.raw = stateTransition.toBuffer().toString('hex')
```
Comment (pshenmic) [inline packages/api/src/utils.js:72]: User fee increase is missing

File: packages/api/src/utils.js [hunk 10]
```diff
@@ -33,23 +35,48 @@ const decodeStateTransition = async (client, base64) => {
...
         dataContractId: documentTransition.getDataContractId().toString(),
-        action: documentTransition.getAction(),
-        revision: documentTransition.getRevision()
+        revision: documentTransition.getRevision(),
+        type: documentTransition.getType(),
+        action: DocumentActionEnum[documentTransition.getAction()]
       }))
 
+      decoded.signature = Buffer.from(stateTransition.getSignature()).toString('hex')
+      decoded.publicKeyId = stateTransition.getSignaturePublicKeyId()
```
Comment (pshenmic) [inline packages/api/src/utils.js:70]: use same field name everywhere

File: packages/api/src/utils.js [hunk 11]
```diff
@@ -33,23 +35,48 @@ const decodeStateTransition = async (client, base64) => {
...
+      decoded.schema = stateTransition.getDataContract().getDocumentSchemas()
+      decoded.signature = Buffer.from(stateTransition.toObject().signature).toString('hex')
+      decoded.signatureKeyId = stateTransition.toObject().signaturePublicKeyId
+      decoded.raw = stateTransition.toBuffer().toString('hex')
 
       break
     }
     case StateTransitionEnum.DOCUMENTS_BATCH: {
       decoded.transitions = stateTransition.getTransitions().map((documentTransition) => ({
         id: documentTransition.getId().toString(),
```
Comment (pshenmic) [inline packages/api/src/utils.js:62]: Different document types have different field set

Here is what is missing:
* Create - data, prefundedVotingBalance, identityContractNonce, documentType
* Replace - data, identityContractNonce, documentType
* Transfer - identityContractNonce, documentType
* UpdatePrice - price, documentType, identityContractNonce
* Purchase - price, identityContractNonce, documentType

**Review by pshenmic — CHANGES_REQUESTED** [review]

File: packages/api/src/utils.js (hunk 5 above)
Comment (owl352) [reply packages/api/src/utils.js:79]: ✅

File: packages/api/README.md (hunk 1 above)
Comment (owl352) [reply packages/api/README.md:61]: Exist

File: packages/api/src/utils.js [hunk 12]
```diff
@@ -88,6 +143,9 @@ const decodeStateTransition = async (client, base64) => {
       decoded.nonce = parseInt(stateTransition.getNonce())
       decoded.outputScript = stateTransition.getOutputScript()?.toString('hex') ?? null
       decoded.coreFeePerByte = stateTransition.getCoreFeePerByte()
+      decoded.signature = stateTransition.getSignature()?.toString('hex') ?? null
```
Comment (pshenmic) [inline packages/api/src/utils.js:146]: Signature Public Key Id is missing
User fee increase

File: packages/api/src/utils.js [hunk 13]
```diff
@@ -88,6 +143,9 @@ const decodeStateTransition = async (client, base64) => {
...
+      decoded.pooling = PoolingEnum[stateTransition.getPooling()]
```
Comment (pshenmic) [inline packages/api/src/utils.js:147]: Identity Nonce is missing

File: packages/api/src/utils.js [hunk 14]
```diff
@@ -88,6 +143,9 @@ const decodeStateTransition = async (client, base64) => {
       decoded.nonce = parseInt(stateTransition.getNonce())
       decoded.outputScript = stateTransition.getOutputScript()?.toString('hex') ?? null
```
Comment (pshenmic) [inline packages/api/src/utils.js:144]: Also add additional outputAddress (you might need to have network type)

File: packages/api/src/utils.js (hunk 5 above)
Comment (pshenmic) [inline packages/api/src/utils.js:79]: Missing AssetLock Proof and funding address information

**Review by pshenmic — CHANGES_REQUESTED** [review]

File: packages/api/src/utils.js (hunk 8 above)
Comment (owl352) [reply packages/api/src/utils.js:56]: 1/2 ✅

File: packages/api/src/utils.js (hunk 6 above)
Comment (owl352) [reply packages/api/src/utils.js:90]: 1/2 ✅

File: packages/api/src/utils.js (hunk 5 above)
Comment (owl352) [reply packages/api/src/utils.js:79]: 1/2 ✅

File: packages/api/src/utils.js (hunk 10 above)
Comment (owl352) [reply packages/api/src/utils.js:70]: ✅

File: packages/api/src/utils.js (hunk 7 above)
Comment (owl352) [reply packages/api/src/utils.js:113]: ✅

[the remark this answers is not in the document]
Comment (owl352) [reply packages/api/src/utils.js:113]: ✅

[the remark this answers is not in the document]
Comment (owl352) [reply packages/api/src/utils.js:113]: ✅

File: packages/api/src/utils.js [hunk 15]
```diff
@@ -59,26 +86,54 @@ const decodeStateTransition = async (client, base64) => {
...
     }
     case StateTransitionEnum.IDENTITY_UPDATE: {
       decoded.identityId = stateTransition.getOwnerId().toString()
       decoded.revision = stateTransition.getRevision()
+      decoded.publicKeysToAdd = stateTransition.getPublicKeysToAdd()
+        .map(key => ({
+          ...key.toJSON(),
+          signature: Buffer.from(key.getSignature()).toString('hex')
+        }))
+      decoded.setPublicKeyIdsToDisable = (stateTransition.setPublicKeyIdsToDisable() ?? []).map(key => key.toJSON())
```
[the remark this answers is not in the document]
Comment (owl352) [reply packages/api/src/utils.js:125]: ✅

File: packages/api/src/utils.js [hunk 16]
```diff
@@ -59,26 +86,54 @@ const decodeStateTransition = async (client, base64) => {
...
+      decoded.signature = stateTransition.getSignature().toString('hex')
```
[the remark this answers is not in the document]
Comment (owl352) [reply packages/api/src/utils.js:126]: ✅

File: packages/api/src/utils.js [hunk 17]
```diff
@@ -59,26 +86,54 @@ const decodeStateTransition = async (client, base64) => {
...
+      decoded.raw = stateTransition.toBuffer().toString('hex')
 
       break
     }
     case StateTransitionEnum.IDENTITY_CREDIT_TRANSFER: {
       decoded.senderId = stateTransition.getIdentityId().toString()
       decoded.recipientId = stateTransition.getRecipientId().toString()
       decoded.amount = stateTransition.getAmount()
+      decoded.signature = stateTransition.getSignature()?.toString('hex') ?? null
```
[the remark this answers is not in the document]
Comment (owl352) [reply packages/api/src/utils.js:135]: ✅

File: packages/api/src/utils.js (hunk 12 above)
Comment (owl352) [reply packages/api/src/utils.js:146]: 1/2 ✅

File: packages/api/src/utils.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (owl352) [reply packages/api/src/utils.js:113]: ✅

File: packages/api/src/utils.js (hunk 11 above)
Comment (owl352) [reply packages/api/src/utils.js:62]: document type already exist

File: packages/api/src/utils.js (hunk 8 above)
Comment (owl352) [reply packages/api/src/utils.js:56]: 2/2 ✅

File: packages/api/src/utils.js (hunk 11 above)
Comment (owl352) [reply packages/api/src/utils.js]: ✅

File: packages/api/src/utils.js [hunk 18]
```diff
@@ -33,61 +37,163 @@ const decodeStateTransition = async (client, base64) => {
...
+      }
+
+      decoded.userFeeIncrease = stateTransition.toObject().userFeeIncrease
+      decoded.identityNonce = Number(stateTransition.getIdentityNonce())
       decoded.dataContractId = stateTransition.getDataContract().getId().toString()
-      decoded.identityId = stateTransition.getOwnerId().toString()
+      decoded.ownerId = stateTransition.getOwnerId().toString()
+      decoded.schema = stateTransition.getDataContract().getDocumentSchemas()
+      decoded.signature = Buffer.from(stateTransition.toObject().signature).toString('hex')
+      decoded.publicKeyId = stateTransition.toObject().signaturePublicKeyId
```
Comment (pshenmic) [inline packages/api/src/utils.js:59]: signaturePublicKeyId

File: packages/api/README.md [hunk 19]
```diff
@@ -951,3 +963,395 @@ Response codes:
...
+        }
+    ]
+}
+```
+```
+{
+    "type": 3,
+    "input": "yfcGcMzUw3RANR5f2qYavnxbQobLD93APi",
+    "assetLockProof": {
+        "instantLock": "AQHlBKj77xbL0VC5XRKhuYg3RIqJTVfYQ4EWUtu6lmdVmAEAAADvzDrsdTKnOK8vC7BZ9uTw3xT3+j7YvrBwM8ZIWlCo9FQPCW6UgYX1XXl8oLqo9TYuDj4CgSStivEwPNTdAAAAhNXZYqwU/zhpaCKxSFLtKe+V9bhm8Gxtn533AC1gBoLP32Z0htf3AdPB9yoKI314GM4XsMZECqOB1HWu/tHmZ8c7JVg0Vp8OMShvEWWXbBTpy3vy5aAe8o5AtyLSc5VE",
```
Comment (pshenmic) [inline packages/api/README.md:1226]: There could be instantSend lock or chainlock, we need to document both cases

File: packages/api/README.md [hunk 20]
```diff
@@ -951,3 +963,395 @@ Response codes:
...
+            "data": "AjJqjBmhxY0w0ULhE9Dd84HZWmMG9WyexNPLjEgjaFsp",
+            "signature": "1f5bb82721b58d92e67db9fb699ab939ccc4a6d5e2e498e80dfb8a3535c13f571923f045e645a898762f8305a4a2218bfedb060f8a8492c48ae9c96247ce17710b"
+        },
+        {
+            "$version": "0",
+            "id": 3,
+            "type": 0,
+            "purpose": 3,
+            "securityLevel": 1,
+            "contractBounds": null,
```
Comment (pshenmic) [inline packages/api/README.md:1213]: add an example with contractBounds

File: packages/api/README.md [hunk 21]
```diff
@@ -951,3 +963,395 @@ Response codes:
...
+```
+{
+    "type": 1,
+    "transitions": [
+        {
+            "id": "2qHj3sdxdKD4G7ZXGCuF6cB6ADYdtPs2BKNx24YcH6gW",
+            "dataContractId": "GWRSAVFMjXx8HpQFaNJMqBV7MBgMK4br5UESsB4S31Ec",
+            "revision": 1,
+            "type": "preorder",
+            "action": "Create",
```
Comment (pshenmic) [inline packages/api/README.md:1147]: lets make it a number

File: packages/api/src/utils.js [hunk 22]
```diff
@@ -1,11 +1,15 @@
...
+const PoolingEnum = require('./enums/PoolingEnum')
+const DocumentActionEnum = require('./enums/DocumentActionEnum')
 const net = require('net')
-const { TCP_CONNECT_TIMEOUT, DPNS_CONTRACT } = require('./constants')
+const { TCP_CONNECT_TIMEOUT, DPNS_CONTRACT, NETWORK } = require('./constants')
 const { base58 } = require('@scure/base')
 const convertToHomographSafeChars = require('dash/build/utils/convertToHomographSafeChars').default
 const Intervals = require('./enums/IntervalsEnum')
 
+const dashcorelib = require('@dashevo/dashcore-lib')
```
Comment (pshenmic) [inline packages/api/src/utils.js:11]: Move to the top

File: packages/api/README.md [hunk 23]
```diff
@@ -58,13 +58,15 @@ Reference:
...
 * [Identity by DPNS](#identity-by-dpns)
+* [Identity Withdrawals](#identity-withdrawals)
 * [Identities](#identities)
 * [Data Contracts by Identity](#data-contracts-by-identity)
 * [Documents by Identity](#documents-by-identity)
 * [Transactions By Identity](#transactions-by-identity)
 * [Transfers by Identity](#transfers-by-identity)
 * [Transactions history](#transactions-history)
 * [Rate](#rate)
+* [Decode StateTransition](#decode-state-transitions)
```
Comment (pshenmic) [inline packages/api/README.md:69]: Decode Raw Transaction

File: packages/api/README.md [hunk 24]
```diff
@@ -951,3 +963,395 @@ Response codes:
...
+    "transitions": [
+        {
+            "id": "2qHj3sdxdKD4G7ZXGCuF6cB6ADYdtPs2BKNx24YcH6gW",
+            "dataContractId": "GWRSAVFMjXx8HpQFaNJMqBV7MBgMK4br5UESsB4S31Ec",
+            "revision": 1,
+            "type": "preorder",
+            "action": "Create",
+            "data": {
+                "saltedDomainHash": "DcKS9AWVE1atKvIokA7JNdUNmyj4SbFUvB6e83whw2g="
+            }
```
Comment (pshenmic) [inline packages/api/README.md:1150]: Add prefundedBalance documentation

File: packages/api/src/utils.js [hunk 25]
```diff
@@ -33,61 +37,163 @@ const decodeStateTransition = async (client, base64) => {
...
       decoded.recipientId = stateTransition.getRecipientId().toString()
       decoded.amount = stateTransition.getAmount()
+      decoded.publicKeyId = stateTransition.toObject().signaturePublicKeyId
+      decoded.signature = stateTransition.getSignature()?.toString('hex') ?? null
+      decoded.raw = stateTransition.toBuffer().toString('hex')
 
       break
     }
     case StateTransitionEnum.IDENTITY_CREDIT_WITHDRAWAL: {
+      decoded.output = dashcorelib.Script(stateTransition.getOutputScript()).toAddress(NETWORK).toString()
```
Comment (pshenmic) [inline packages/api/src/utils.js:185]: outputAddress

File: packages/api/src/utils.js [hunk 26]
```diff
@@ -33,61 +37,163 @@ const decodeStateTransition = async (client, base64) => {
...
+        signature: Buffer.from(key.getSignature()).toString('hex')
+      }))
 
       break
     }
     case StateTransitionEnum.IDENTITY_TOP_UP: {
       const assetLockProof = stateTransition.getAssetLockProof()
       const output = assetLockProof.getOutput()
 
+      decoded.input = dashcorelib.Script(assetLockProof.getOutput().script).toAddress(NETWORK).toString()
```
Comment (pshenmic) [inline packages/api/src/utils.js:118]: fundingAddress

**Review by pshenmic — CHANGES_REQUESTED** [review]

File: packages/api/README.md (hunk 21 above)
Comment (owl352) [reply packages/api/README.md:1147]: ✅

File: packages/api/src/utils.js (hunk 22 above)
Comment (owl352) [reply packages/api/src/utils.js:11]: ✅

File: packages/api/README.md (hunk 23 above)
Comment (owl352) [reply packages/api/README.md:69]: ✅

File: packages/api/src/utils.js (hunk 26 above)
Comment (owl352) [reply packages/api/src/utils.js:118]: ✅

File: packages/api/src/utils.js (hunk 25 above)
Comment (owl352) [reply packages/api/src/utils.js:185]: ✅

File: packages/api/README.md (hunk 24 above)
Comment (owl352) [reply packages/api/README.md:1150]: ✅

File: packages/api/README.md (hunk 19 above)
Comment (owl352) [reply packages/api/README.md:1226]: ✅

**Pull request merged**